packages/hooks_riverpod/test/provider_listener_hook_test.dart

```diff
@@ -0,0 +1,279 @@
+    );
+
+    key.currentContext.read(provider).state = -1;
+    await tester.pump();
+    verify(onChange(true)).called(1);
+
+    key.currentContext.read(provider).state = 2;
+    await tester.pump();
+    verify(onChange(false)).called(1);
+  });
+
+  testWidgets('can mark parents as dirty during onChange', (tester) async {
+    final container = ProviderContainer();
+    final provider = StateProvider((ref) => 0);
+    final onChange = ListenerMock<int>();
+
+    await tester.pumpWidget(
+      StatefulBuilder(builder: (context, setState) {
+        return UncontrolledProviderScope(
+          container: container,
+          child: HookBuilder(
+            builder: (context) {
+              useProviderListener<StateController<int>>(
+                  provider, (_, value) => onChange(value.state));
+              return Container();
+            },
+          ),
+        );
+      }),
+    );
+
+    verifyZeroInteractions(onChange);
+
+    container.read(provider).state++;
+    await Future<void>.value();
+  });
+
+  testWidgets('calls onChange at most once per frame', (tester) async {
+    final provider = StateProvider((ref) => 0);
+    final onChange = ListenerMock<int>();
+    final container = ProviderContainer();
+
+    await tester.pumpWidget(
+      UncontrolledProviderScope(
+        container: container,
+        child: HookBuilder(builder: (context) {
+          useProviderListener<StateController<int>>(
+              provider, (_, value) => onChange(value.state));
+          return Container();
+        }),
+      ),
+    );
+    verifyZeroInteractions(onChange);
+
+    container.read(provider).state++;
+    container.read(provider).state++;
+    container.read(provider).state++;
+    await Future<void>.value();
+
+    verifyOnly(onChange, onChange(3));
+  });
+
+  testWidgets('supports Changing the ProviderContainer', (tester) async {
+    final provider = Provider((ref) => 0);
+    final onChange = ListenerMock<int>();
+    final container = ProviderContainer(overrides: [
+      provider.overrideWithValue(0),
+    ]);
+    final container2 = ProviderContainer(overrides: [
+      provider.overrideWithValue(0),
+    ]);
+
+    await tester.pumpWidget(
+      UncontrolledProviderScope(
+        container: container,
+        child: HookBuilder(builder: (context) {
+          useProviderListener<int>(provider, (_, value) => onChange(value));
+          return Container();
+        }),
+      ),
+    );
+
+    await tester.pumpWidget(
+      UncontrolledProviderScope(
+        container: container2,
+        child: HookBuilder(builder: (context) {
+          useProviderListener<int>(provider, (_, value) => onChange(value));
+          return Container();
+        }),
+      ),
+    );
+
+    container.updateOverrides([
+      provider.overrideWithValue(21),
+    ]);
+    container2.updateOverrides([
+      provider.overrideWithValue(42),
+    ]);
+
+    await Future<void>.value();
+
+    verifyOnly(onChange, onChange(42));
```

Comment:
Can we have two `onChange` (one per container), and verify that the onChange called is the one from the most up-to-date container?

That test could pass even without the related code